grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/DeferredClientCall.java

```diff
@@ -0,0 +1,156 @@
+                return null;
+            }
+            if (!(delegate instanceof ArmeriaClientCall)) {
+                // TODO
+                return null;
+            }
+            lock.lock();
+            try {
+                this.delegate = (ArmeriaClientCall<I, O>) delegate;
+            } finally {
+                lock.unlock();
+            }
+            drainPendingTasks();
+            return null;
+        });
+    }
+
+    @Override
+    public void start(Listener<O> responseListener, Metadata headers) {
+        lock.lock();
+        try {
+            if (delegate != null) {
+                delegate.start(responseListener, headers);
+            } else {
+                pendingTasks.add(() -> delegate.start(responseListener, headers));
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void request(int numMessages) {
+        lock.lock();
+        try {
+            if (delegate != null) {
+                delegate.request(numMessages);
+            } else {
+                pendingTasks.add(() -> delegate.request(numMessages));
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void cancel(@Nullable String message, @Nullable Throwable cause) {
+        lock.lock();
+        try {
+            if (delegate != null) {
+                delegate.cancel(message, cause);
+            } else {
+                pendingTasks.add(() -> delegate.cancel(message, cause));
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void halfClose() {
+        lock.lock();
+        try {
+            if (delegate != null) {
+                delegate.halfClose();
+            } else {
+                pendingTasks.add(this::halfClose);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void sendMessage(I message) {
+        lock.lock();
+        try {
+            if (delegate != null) {
+                delegate.sendMessage(message);
+            } else {
+                pendingTasks.add(() -> delegate.sendMessage(message));
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void drainPendingTasks() {
+        List<Runnable> runnables = new ArrayList<>();
+        while (true) {
+            lock.lock();
+            try {
+                if (pendingTasks.isEmpty()) {
+                    break;
+                }
+            } finally {
+                lock.unlock();
+            }
+            lock.lock();
+            try {
+                final List<Runnable> tmp = runnables;
+                runnables = pendingTasks;
+                pendingTasks = tmp;
+            } finally {
+                lock.unlock();
+            }
+            for (Runnable runnable : runnables) {
+                delegate.ctx().makeContextAware(runnable).run();
+            }
```

Comment:
- We should use `ArrayDeque` and `poll()` instead of iteration? https://github.com/line/armeria/pull/4729#discussion_r1150093517
- It is more efficient to invoke `ctx.push()` directly instead of wrapping `Runnable`.
```java
for (;;) {
    Runnable runnable = queue.poll();
    if (runnable == null) {
        break;
    }
    try (SafeCloseable ignored = delegate.ctx.push()) { 
        runnable.run();
    } catch(Throwable ex) {
        delegate.cancel(...);
        break;
    }
}
```

Reply:
Just curious.
Why is `for(;;)` preferred in this project?

Reply:
> Why is `for(;;)` preferred in this project?

It was probably from my habit? `while (true)` implies one boolean comparison although it will be optimized away. `for (;;)` is a pure loop in contrast. I wish the language has `loop { ... }` like Swift has.